services/ExtraActivityManager.php

```diff
@@ -0,0 +1,340 @@
+        $this->wiki = $wiki;
+    }
+
+    /**
+     * Save a Extra-activity
+     * @param array $data $_POST data to save
+     * @return bool false if error
+     */
+    public function saveExtraActivity(array $data): bool
+    {
+        $debug = ($this->wiki->GetConfigValue('debug')=='yes');
+        // === START validate data ====
+        $error = (empty($data['title'])
+            || empty($data['bf_date_debut_evenement'])
+            || empty($data['bf_date_fin_evenement'])
+            || empty($data['course'])
+            || !isset($data['registeredLearnerNames'])
+            || count($data['registeredLearnerNames']) == 0) ;
+        if ($error) {
+            if ($debug) {
+                $output = 'Errors in '. get_class($this) . ' :<br>' ;
+                $output .= (empty($data['title'])) ? 'empty($data[\'title\'])<br>' : '' ;
+                $output .= (empty($data['bf_date_debut_evenement'])) ? 'empty($data[\'bf_date_debut_evenement\'])<br>' : '' ;
+                $output .= (empty($data['bf_date_fin_evenement'])) ? 'empty($data[\'bf_date_fin_evenement\'])<br>' : '' ;
+                $output .= (empty($data['course'])) ? 'empty($data[\'course\'])<br>' : '' ;
+                $output .= (!isset($data['registeredLearnerNames'])) ? 'not isset($data[\'registeredLearnerNames\'])<br>' : '' ;
+                $output .= (isset($data['registeredLearnerNames']) && count($data['registeredLearnerNames']) == 0) ? 'count($data[\'registeredLearnerNames\']) == 0<br>' : '' ;
+                echo $output;
+            }
+            return false ;
+        }
+        // === END validate data ====
+
+        // === START check tag ====
+        if (!empty($data['tag'])) {
+            $extraActivities = $this->getExtraActivityLogsFromLike('%"tag":"' . $data['tag'] . '"%');
+            if (!$extraActivities->has($data['tag'])) {
+                if ($debug) {
+                    echo 'Errors in '. get_class($this) . ' : $data[\'tag\'] defined but not existing in triples' .'<br>' ;
+                }
+                return false;
+            } else {
+                $oldExtraActivity = $extraActivities->get($data['tag']);
+            }
+        } else {
+            // define new Tag
+            $tmptag = genere_nom_wiki($data['title'], 1);
+        
+            // check if tag is a pageName or already exists for this course
+            // get all extra-activities
+            $extraActivities = $this->getExtraActivityLogsFromLike('');
+
+            $i = 1 ;
+            do {
+                $tag = genere_nom_wiki($tmptag, $i);
+                ++$i;
+            } while ($i < 1000 && $extraActivities->has($tag)) ;
+            if ($extraActivities->has($tag)) {
+                if ($debug) {
+                    echo 'Errors in '. get_class($this) . ' : genere_nom_wiki does not work' .'<br>' ;
+                }
+                return false;
+            } else {
+                $data['tag'] = $tag ;
+            }
```

Comment:
De ce que je comprends si le tag n'est pas renseigné (peut aussi être une url aussi d'après les specs si on parle bien du même truc), tu vas le renseigner en générant l'identifiant à partir du titre (pour la génération de l'identifiant, il ne faudrait pas réécrire autant ici, ce n'est pas la responsabilité de cette classe).
Ce n'est pas ce que j'avais compris des specs. Il suffit à mon avis de simplement laisser l'utilisateur renseigner le tag. Par contre, ce serait mieux de vérifier que le tag existe. 
C'est pour cela que ça serait certainement plus simple de mettre un checkboxfiche sur le formulaire des émargements. Et si c'est vraiment important d'avoir une url, on peut laisser aussi un champ texte. Ce qui n'empêche pas d'avoir au niveau du triple qu'un seul champ si on préfère. A voir avec @mrflos pour la solution retenue.

Reply:
Attention @acheype je pense qu'il y a confusion.

Pour fonctionner, j'ai besoin d'un identifiant unique de l'extra-activité, et je ne souhaite pas que cet identifiant soit aussi  celui d'une page.

c'est juste ceci.

Reply:
oui j'ai confondu deux trucs.
ok pour le tag unique pour l'extra activity, mais faudrait améliore ce code. J'ai un exemple dans login-sso où j'avais réécris cette fonction en permettant de changer la condition qui permet d'itérer le numéro. Tu peux y jeter un oeil ici : https://github.com/YesWiki/yeswiki-extension-login-sso/blob/21bd89d68d6cc13ac8709b2c42a0cbc982e1eba8/libs/login-sso.lib.php#L144

ça vaudrait le coup d'intégrer ce code au cœur, voire de l'améliorer (que j'avais prévu de faire le jour où je bosse à nouveau sur les différents modules login)

Reply:
Désolé par le temps, je fais avec l'existant dans le coeur.

Reply:
L'autre soucis était que pour générer le tag ça continuait de tester si une page n'avait pas ce nom là.
Je t'ai fais une PR pour que tu ais plus qu'à lancer la fonction : https://github.com/YesWiki/yeswiki/pull/717

Reply:
J'espère que tu ne m'en voudras pas @acheype mais vu que ça marche pour le moment, je préfère remettre à après le 12/04 l'intégration et le test de ce que tu proposes car ça ne me semble pas prioritaire (la fonction étant rarement appelée, il n'est pas nécessaire d'optimiser de suite son fonctionnement).

Reply:
du moment que c'est fait par la suite, pas de soucis.
il y a toutefois un bug actuellement, si tu nommes ton triple avec le tag 'Essai' et qu'une page 'Essai' existe déjà, ça te rend un tag 'Essai2' bien que 'Essai' soit disponible pour le triple.

Reply:
Ça n'est pas un bug, c'est ce que je veux faire.

Reply:
Pour moi, il faut avoir un tag unique pour les triples et non unique pour les triples ET les pages.

Reply:
c'est aussi une bonne idée, à rediscuter après le 12/04

Reply:
oui pas d'urgence, y'a rien de critique.
C'est ce qu'en tout cas ce que te permet de faire la PR que j'ai écrite avec :
```
generateTagNameFromCondition($nom, function($tag){ 
   // requête dans la base de données pour voir si le triple existe déjà avec ce tag
});
```

Reply:
Je ne ferai pas le changement car:
1. je ne suis pas d'accord pour utiliser la PR https://github.com/YesWiki/yeswiki/pull/717 que je trouve plus compliquée sur la solution actuelle.
2. ce bout de code est rarement appelé (juste pour la création d'une extra activité) du coup, je ne vois pas trop l'intérêt d'automatiser le temps de traitement vu que ça fonctionne.